File: linkedList.js
class Node {
  constructor(val) {
		this.val = val;
		this.next = null;
	}
}

class LinkedList {
	constructor() {
		this.head = null;
		this.tail = null;
		this.length = 0;
	}

	insertHead(val) {
		const node = new Node(val);

		if (!this.head) {
			this.head = node;
			this.tail = node;
		} else {
			node.next = this.head;
			this.head = node;
		}

		this.length++;
	}

	insertTail(val) {
		const node = new Node(val);

		if (!this.head) {
			this.head = node;
		}else {
			this.tail.next = node;
		}

		this.tail = node;
		this.length++;
	}

	removeHead() {
		if (!this.head) return null;

		const head = this.head.val;

		if (this.length === 1) {
			this.head = null;
			this.tail = null;
		} else {
			this.head = this.head.next;
		}

		this.length--;
		return head;
	}

	removeTail() {
		if (!this.head) return null;

		const tail = this.tail.val;

		if (this.length === 1) {
			this.head = null;
			this.tail = null;
		} else {
			let curr = this.head;
			while (curr.next) {
				if (curr.next === this.tail) {
					this.tail = curr;
					curr.next = null;
				} else {
					curr = curr.next;
				}
			}
		}

		this.length--;
		return tail;
	}

	containsNode(val) {
		let curr = this.head;

		while (curr) {
			if (curr.val === val) return true;
			curr = curr.next;
		}

		return false;
	}

}

File: linked_list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def add_head(self, val):
        node = Node(val)

        if self.length: node.next = self.head
        else: self.tail = node

        self.head = node
        self.length += 1

    def add_tail(self, val):
        node = Node(val)

        if self.length:
            self.tail.next = node
        else: self.head = node

        self.tail = node
        self.length += 1

    def remove_head(self):

        if self.head: head = self.head.val
        else: return None

        if self.length == 1: self.tail = None
        self.head = self.head.next

        self.length -= 1
        return head

    def remove_tail(self):
        if self.tail: tail = self.tail.val
        else: return None

        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            curr = self.head

            while curr:
                if curr.next is self.tail:
                    curr.next = None
                    self.tail = curr

                curr = curr.next

        self.length -= 1
        return tail

    def contains_node(self, val):
        if not self.length: return False

        curr = self.head

        while curr:
            if curr.val == val: return True
            curr = curr.next

        return False


l = LinkedList()
l.add_head(10)
l.add_head(8)
l.add_tail(13)
l.add_head(5)
l.add_tail(15)
l.remove_head()
l.remove_tail()
l.contains_node(5)
l.contains_node(10)

# 5, 8, 10, 13, 15